src/main/navigation/navigation_rover_boat.c

```diff
@@ -71,11 +71,11 @@ void applyRoverBoatPositionController(timeUs_t currentTimeUs)
     static timeUs_t previousTimePositionUpdate;         // Occurs @ GPS update rate
     static timeUs_t previousTimeUpdate;                 // Occurs @ looptime rate
 
-    const timeDelta_t deltaMicros = currentTimeUs - previousTimeUpdate;
+    const timeDeltaLarge_t deltaMicros = currentTimeUs - previousTimeUpdate;
```

Comment:
I don't see how this fixes the problem. We can still get negative `deltaMicros`, just much later. I would like to avoid 64-bit math as it's expensive in terms of CPU cycles. A better mitigation would be to ensure `deltaMicros` is positive.

Reply:
No worries, I will explain my thought process a little bit better.

---

> We can still get negative `deltaMicros`

I disagree. Assuming signed type negative `deltaMicros` can be a result of:
1. `currentTimeUs < previousTimeUpdate` - this case will not happen due to semantics of `previousTimeUpdate` with a simple assumption that `currentTimeUs` is non-decreasing
2. `TYPE_MAX < (currentTimeUs - previousTimeUpdate) < (2 * TYPE_MAX)` (for simplicity we are considering only the first interval) - this is the case that can happen and happened to me. With 32-bit type the problem pops up in the interval [~36&nbsp;mins,&nbsp;~72 mins], which is an interval that can easily be seen in a perfectly normal use case. On the other hand 64-bit type gives us interval of failure [~300000&nbsp;years,&nbsp;~600000 years] and I don't believe it is sane to talk about the craft being on for 300000&nbsp;years. (fun fact: 300000&nbsp;years ago Homo Sapiens started to apppear on Earth)

Therefore I believe we cannot get negative `deltaMicros`.

---

> avoid 64-bit math as it's expensive in terms of CPU cycles

Note that any function taking deltaMicros into any serious calculation (like `update2DPositionHeadingController` or `updateAltitudeVelocityController_MC`) still uses `timeDelta_t`, so performance there is not changed at all. Changes from 32 to 64-bit variables are very selective and after making sure it fits the delta is converted to 32-bit (implicitly, because of function's parameter type). And the check if it fits does not even have to be added. It is already being done, because delta is always checked if it is not too big and not too big value is asserted to fit in the 32-bit type.
To summarize: only a few more cpu cycles will be used during the first comparison, but this amount is small and necessary. All&nbsp;serious calculations are left unchanged.

---

> A better mitigation would be to ensure `deltaMicros` is positive.

I disagree. After ~72 mins the delta held in 32-bit type would indicate a positive value again, but this value will be wrong (because it wraped around) and can falsely show for example 1us when it should show 4294967297us. This can lead to wrongly satisfying all checks and proceeding with calculations without resetting appropriate controllers after a ~72min delay. So&nbsp;I&nbsp;believe this one would be a solution that does not actually solve the problem, just pushes it by half an hour.

Reply:
Fair enough, you got me convinced :smile: